libs/sql-schema-describer/src/mssql.rs

```diff
@@ -0,0 +1,570 @@
+
+        let mut columns = self.get_all_columns(schema).await;
+        let mut indexes = self.get_all_indices(schema).await;
+        let mut foreign_keys = self.get_foreign_keys(schema).await;
+
+        let table_names = self.get_table_names(schema).await;
+        let mut tables = Vec::with_capacity(table_names.len());
+
+        for table_name in table_names {
+            let table = self.get_table(&table_name, &mut columns, &mut indexes, &mut foreign_keys);
+            tables.push(table);
+        }
+
+        Ok(SqlSchema {
+            tables,
+            enums: vec![],
+            sequences: vec![],
+        })
+    }
+}
+
+impl SqlSchemaDescriber {
+    pub fn new(conn: Arc<dyn Queryable + Send + Sync + 'static>) -> Self {
+        Self { conn }
+    }
+
+    async fn get_databases(&self) -> Vec<String> {
+        debug!("Getting databases");
+
+        let sql = "SELECT name FROM master.sys.databases";
+        let rows = self.conn.query_raw(sql, &[]).await.expect("get schema names");
+
+        let names = rows
+            .into_iter()
+            .map(|row| {
+                row.get("name")
+                    .and_then(|x| x.to_string())
+                    .expect("convert schema names")
+            })
+            .collect();
+
+        debug!("Found schema names: {:?}", names);
+
+        names
+    }
+
+    async fn get_table_names(&self, schema: &str) -> Vec<String> {
+        debug!("Getting table names");
+
+        let select = r#"
+            SELECT table_name
+            FROM information_schema.tables
+            WHERE table_schema = @P1
+            AND table_type = 'BASE TABLE'
+            ORDER BY table_name ASC
+        "#;
+
+        let rows = self
+            .conn
+            .query_raw(select, &[schema.into()])
+            .await
+            .expect("get table names");
+
+        let names = rows
+            .into_iter()
+            .map(|row| {
+                row.get("table_name")
+                    .and_then(|x| x.to_string())
+                    .expect("get table name")
+            })
+            .collect();
+
+        debug!("Found table names: {:?}", names);
+
+        names
+    }
+
+    async fn get_size(&self, schema: &str) -> usize {
+        debug!("Getting db size");
+
+        let sql = r#"
+            SELECT SUM(alloUni.used_pages) * 8000 AS size
+            FROM sys.tables sysTab
+            INNER JOIN sys.indexes ind
+            ON sysTab.OBJECT_ID = ind.OBJECT_ID
+            AND ind.Index_ID <= 1
+            INNER JOIN sys.partitions parti
+                ON ind.OBJECT_ID = parti.OBJECT_ID
+                AND ind.index_id = parti.index_id
+            INNER JOIN sys.allocation_units alloUni
+                ON parti.partition_id = alloUni.container_id
+            WHERE SCHEMA_NAME(sysTab.SCHEMA_ID) == @P1
+            AND sysTab.is_ms_shipped = 0
+            AND ind.OBJECT_ID > 255
+            AND parti.rows > 0
+        "#;
```

Comment:
woot

Reply:
I know...

Reply:
So the last filters should take out all Microsoft-provided system tables, counting the size of the actual user data.